Digram

aalexandrasilva opened the pull request:
Added the following implementation:
1) Finding all digram occurrences in a graph (any two edges connected to a common node), by using three main cases:
- Common node is object in one, and subject in other: n1 e1 **n2** -> **n2** e2 n3
- Common node is object in both: n1 e1 **n2** -> n3 e2 **n2**
- Common node is subject in both: **n2** e1 n1 -> **n2** e2 n3

2) Sorting the digrams by most to least frequent and discarding the overlapping occurrences

TortugaAttack commented on src/main/java/org/dice_group/grp/grammar/digram/Digram.java:
```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
```
instead you can ask if(obj instanceof Digram) 
In this case it does not matter though

TortugaAttack commented on src/main/java/org/dice_group/grp/grammar/digram/Digram.java:
```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
+				return false;
+		} else if (!edgeLabel2.equals(other.edgeLabel2))
+			return false;
+		if (external == null) {
```
externals are not allowed to be empty/null according to the paper. So this case can be removed

MichaelRoeder replied:
Well, following the idea of defensive programming, you may simply want to throw an exception with an error message indicating that this case is not allowed :wink:

TortugaAttack replied:
true.

aalexandrasilva replied:
my bad, I used the auto-generated equals by the IDE. But I changed it to incorporate your reviews :)

TortugaAttack commented on src/main/java/org/dice_group/grp/grammar/digram/Digram.java:
```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```
all the edgeLabelX == null can be removed, if an edgeLabel is null there is something wrong. :)

MichaelRoeder replied:
Again, you could also throw an exception. Or use [`Objects.requireNonNull()`](https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#requireNonNull(T,%20java.lang.String))

TortugaAttack commented on src/test/java/org/dice_group/grp/grammar/digram/DigramTest.java:
```diff
@@ -0,0 +1,43 @@
+public class DigramTest {
+
+	@Test
+	public void test() {
```
should have an assertion that everything went correct.

TortugaAttack commented on src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java:
```diff
@@ -47,14 +68,275 @@
+	 */
+	private static boolean isNodeExternal(String nodeString, int diff, Model graph) {
+		StringBuilder query = new StringBuilder();
+		query.append("SELECT (COUNT(*)-").append(diff).append(" AS ?total) WHERE { { ")
```
Shouldnt this be an ASK query with MINUS, as we discussed?

TortugaAttack commented on src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java:
```diff
@@ -47,14 +68,275 @@
+	 * @param node
+	 * @return
+	 */
+	private static String getNodeName(RDFNode node) {
```
There is a class called JenaNodeFormatter.format() in the HDT library (included) this does exactly that :)

TortugaAttack commented on src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java:
```diff
@@ -47,14 +68,275 @@
+	 * @param sparqlQuery
+	 * @return
+	 */
+	public static QueryExecution queryModel(Model graph, String sparqlQuery) {
```
maybe queryModel, getNodeName and selectModel in its own RDFHelper class. So it is a little bit more lucid

TortugaAttack commented on src/main/java/org/dice_group/grp/grammar/digram/DigramHelper.java:
```diff
@@ -47,14 +68,275 @@
+	 * @param occurrences
+	 * @return
+	 */
+	public static Map<Digram, Set<DigramOccurence>> findNonOverOccurrences(Set <DigramOccurence> occurrences){
```
I like the worst case O(n^2) best case O(n) greedy approach. looks good to me.
However this is heuristic, we might have to come back to check if we can get another approach implemented in the same worst case time which is not heuristic. 
Then we can compare the two approaches and check if the greedy algorithm is sufficient in most cases.

TortugaAttack replied:
Acutally i have one written on paper, so i can do that.

aalexandrasilva commented:
> Pull Request looks good :)
> 
> Minor things however as stated
> Especially the one with the MINUS query and the test assertion.

I have changed the code, the .equals() method has been rectified and as we've discussed the minus query wasn't possible with Jena. 

I've also added new assertions in the test cases :)

TortugaAttack commented:
Thanks :)
